src/js/map/WeightsGrid.js

```diff
@@ -1,62 +1,60 @@
-
 import Grid from './Grid'
+import TerrainGrid from './TerrainGrid'
+import TowerGrid from './TowerGrid'
 
 export default class WeightsGrid extends Grid {
   constructor(tilesWide, tilesHigh) {
-    // this.grid = new Grid(tilesWide, tilesHigh)
     super(tilesWide, tilesHigh, 1)
+    this.terrain = new TerrainGrid(tilesWide, tilesHigh)
+    this.tower = new TowerGrid(tilesWide, tilesHigh)
+    this.compute()
   }
 
-  randomize(wallProbability = 0.1) {
-    for (let i = 0; i < this.tilesWide; i++) {
-      for (let j = 0; j < this.tilesHigh; j++) {
-        let random = Math.random()
-        if (random < wallProbability) {
-          this.set(i, j, 0)
-        } else {
-          this.set(i, j, 1)
-        }
-      }
-    }
+  compute() {
+    this.recalculate({x: 0, y: 0}, this.tilesWide, this.tilesHigh)
   }
 
-  /*
-   * Adds an obstacle onto the weights grid. Simply sets relevant values to 0.
-   * Takes a position, a width, and a height.
-   */
-  addObstacle(gridLocation, gridWidth, gridHeight) {
-    for (let x = gridLocation.x; x < gridLocation.x + gridWidth; x++) {
-      for (let y = gridLocation.y; y < gridLocation.y + gridHeight; y++) {
-        this.set(x, y, 0)
-      }
-    }
+  reset() {
+    super.reset()
+    this.terrain.reset()
+    this.tower.reset()
+    this.compute()
   }
 
-  isAreaFree(gridLocation, gridWidth, gridHeight) {
-    if ([gridLocation, gridWidth, gridHeight].indexOf(undefined) !== -1) {
-      throw TypeError('Must supply real values to isAreaFree().')
-    }
+  addTerrainObstacle(gridLocation, gridWidth, gridHeight) {
+    this.terrain.addObstacle(gridLocation, gridWidth, gridHeight)
+    this.recalculate(gridLocation, gridWidth, gridHeight)
+  }
+
+  addTowerObstacle(gridLocation, gridWidth, gridHeight) {
+    this.tower.addObstacle(gridLocation, gridWidth, gridHeight)
+    this.recalculate(gridLocation, gridWidth, gridHeight)
+  }
+
+  recalculate(gridLocation, gridWidth, gridHeight) {
     for (let x = gridLocation.x; x < gridLocation.x + gridWidth; x++) {
```

Comment:
This method should maybe handle the case where you give it a gridWidth or gridHeight that is too big - eg. if gridLocation is in the middle, and you give the normal width/height, then this loop will go off the map. I'm not sure how it would currently behave in that situation.

Reply:
Good idea! I think I just slapped it together and the thought did cross (and disappear from) my mind.

Reply:
I think all the methods (addTowerObstacle, addTerrainObstacle, recalculate) all lead to a set(x, y, value) call, which returns false and does nothing else if !coordinateIsValid(x, y). Would we still want to deal with that before it reaches that point (eg. before it iterates)?